<a href="https://colab.research.google.com/github/archisha13/Deep-Learning/blob/main/B072_DL_LAB2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>



>              NAME: ARCHISHA SRIVASTAVA 



>              DL_LAB2

>              ROLL NUM: B072





**AIM:**


To determine error/loss for feed forward network with three layers


In [ ]:
import numpy as np
import pandas as pd
import tensorflow as tf
# tensorflow is used to build the network
import matplotlib as plt

In [ ]:
#reading the dataset
df = pd.read_csv("/content/iris_data.csv")

In [ ]:
df.head()

sepal_length  sepal_width  petal_length  petal_width      species
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

In [ ]:
df.shape
# this data frame has 150 rows and 5 columns

(150, 5)

Whenever numerical values are given as a part of the feature we normalise the values

In [ ]:
#creating a new datset containing
df1 =pd.get_dummies(df["species"])

In [ ]:
df1.head(5)

Iris-setosa  Iris-versicolor  Iris-virginica
0            1                0               0
1            1                0               0
2            1                0               0
3            1                0               0
4            1                0               0

1 0 0 represents Iris setosa

In [ ]:
df['species'].unique()
# we can see there are  3 types of flowers 

array(['Iris-setosa', 'Iris-versicolor', 'Iris-virginica'], dtype=object)

In [ ]:
from sklearn.preprocessing import StandardScaler
std_scalar=StandardScaler()
#applying standard scaler on each column independently 


In [ ]:
''' mean and std deviation of the column ...for each column for each value in that column [ x=(x-mean)/std ]
mean will be 0 and std dev =1'''
df.iloc[ : ,[0,1,2,3]]=std_scalar.fit_transform(df.iloc[:,[0,1,2,3]])

Encoding the categorical variable 

In [ ]:
df.head()

sepal_length  sepal_width  petal_length  petal_width      species
0     -0.900681     1.032057     -1.341272    -1.312977  Iris-setosa
1     -1.143017    -0.124958     -1.341272    -1.312977  Iris-setosa
2     -1.385353     0.337848     -1.398138    -1.312977  Iris-setosa
3     -1.506521     0.106445     -1.284407    -1.312977  Iris-setosa
4     -1.021849     1.263460     -1.341272    -1.312977  Iris-setosa

In [ ]:
df.drop('species',axis =1, inplace =True)
#inplace of original datset the change happends


In [ ]:
df.head()

sepal_length  sepal_width  petal_length  petal_width
0     -0.900681     1.032057     -1.341272    -1.312977
1     -1.143017    -0.124958     -1.341272    -1.312977
2     -1.385353     0.337848     -1.398138    -1.312977
3     -1.506521     0.106445     -1.284407    -1.312977
4     -1.021849     1.263460     -1.341272    -1.312977

In [ ]:
df_final= pd.concat([df,df1],axis=1)
# joining the two datsets

In [ ]:
df_final.head()

sepal_length  sepal_width  petal_length  petal_width  Iris-setosa  \
0     -0.900681     1.032057     -1.341272    -1.312977            1   
1     -1.143017    -0.124958     -1.341272    -1.312977            1   
2     -1.385353     0.337848     -1.398138    -1.312977            1   
3     -1.506521     0.106445     -1.284407    -1.312977            1   
4     -1.021849     1.263460     -1.341272    -1.312977            1   

   Iris-versicolor  Iris-virginica  
0                0               0  
1                0               0  
2                0               0  
3                0               0  
4                0               0

In [ ]:
#there are four nodes in the input layer since 4 features
#there will be 3 nodes in output layer since 3 types of species
#we will take 2 nodes in the hidden layer 

In [ ]:
# w_i1_h1 is the weight from i1 to h1
# w_i_h1 is the weight from all the nodes in input layer to h1

''' generating 4 random values to h1 from node/neuron i1'''
# seed is used to fix the first value
np.random.seed(42)
# each node in hidden layer is recieving 4 inputs from i
w_i_h1 = np.random.rand(4,1)
w_i_h2 = np.random.rand(4,1)
w_i_h3 = np.random.rand(4,1)
# each node in output layer is recieving 3 inputs from h(hidden layer)
w_h_o1 = np.random.rand(3,1)
w_h_o2 = np.random.rand(3,1)
w_h_o3 = np.random.rand(3,1)
#b1 is the bias for hidden layer neuron 1
b1 = np.random.rand(1)
b2 = np.random.rand(1)
#adding new node in hidden layer
b3 = np.random.rand(1)

In [ ]:
print(w_i_h1)

[[0.37454012]
 [0.95071431]
 [0.73199394]
 [0.59865848]]


In [ ]:
#sigmoid function
def sig(x):
  sig_x= 1/(1+np.exp(-x))
  return sig_x

In [ ]:
input = df_final.iloc[ : ,[0,1,2,3]]

#taking all the rows and first four columns

Z2_1====>2nd layer ---1 st node

In [ ]:
#y = w.x + b...where x is the input
z2_1= np.dot(input,w_i_h1)+b1
z2_2= np.dot(input,w_i_h2)+b2
z2_3= np.dot(input,w_i_h3)+b3

#input layer for 2nd layer(hidden layer) is sigmoid(y)
h2_1= sig(z2_1)
h2_2= sig(z2_2)
h2_3= sig(z2_3)

In [ ]:
h_1 = np.append(h2_1,h2_2,axis=1)
h = np.append(h_1,h2_3,axis=1)


In [ ]:
o3_1 = np.dot(h,w_h_o1)
o3_2 = np.dot(h,w_h_o2)
o3_3= np.dot(h,w_h_o3)

In [ ]:
o1 = sig(o3_1)
o2 = sig(o3_2)
o3 = sig(o3_3)
#output layer node values

In [ ]:
#print(o1)

In [ ]:
df_final

sepal_length  sepal_width  petal_length  petal_width  Iris-setosa  \
0       -0.900681     1.032057     -1.341272    -1.312977            1   
1       -1.143017    -0.124958     -1.341272    -1.312977            1   
2       -1.385353     0.337848     -1.398138    -1.312977            1   
3       -1.506521     0.106445     -1.284407    -1.312977            1   
4       -1.021849     1.263460     -1.341272    -1.312977            1   
..            ...          ...           ...          ...          ...   
145      1.038005    -0.124958      0.819624     1.447956            0   
146      0.553333    -1.281972      0.705893     0.922064            0   
147      0.795669    -0.124958      0.819624     1.053537            0   
148      0.432165     0.800654      0.933356     1.447956            0   
149      0.068662    -0.124958      0.762759     0.790591            0   

     Iris-versicolor  Iris-virginica  
0                  0               0  
1                  0               0  
2                  0               0  
3                  0               0  
4                  0               0  
..               ...             ...  
145                0               1  
146                0               1  
147                0               1  
148                0               1  
149                0               1  

[150 rows x 7 columns]

In [ ]:
f1= np.array(df_final.iloc[: ,4])#4 is the column num of Iris setosa
f2= np.array(df_final.iloc[: ,5])# 5 is the column num of Iris-versicolor
f3= np.array(df_final.iloc[: ,6])# 5 is the column num of Iris-virginica

In [ ]:
f1[:3]

array([1, 1, 1], dtype=uint8)

In [ ]:
f1 = f1.reshape((-1,1))
f2 = f2.reshape((-1,1))
f3 = f3.reshape((-1,1))

In [ ]:
f1[:3]

array([[1],
       [1],
       [1]], dtype=uint8)

In [ ]:
e1= f1-o1
e2= f2-o2
e3= f3-o3

In [ ]:
error = np.square(e1)+np.square(e2)+np.square(e3)
error[:5]

array([[0.86144812],
       [0.81064944],
       [0.81740181],
       [0.80820476],
       [0.86811849]])

In [ ]:
total_error= np.sum(error)
total_error

147.44942573200146

**CONCLUSION:**


> For the given Irius dataset Neural network wihb 3 layeras is built


*   Hidden layer has two neurons and sigmoid is used as an activation function

*   For the above architecture with forward propagation , the error/loss is**144.82**for 2 nodes/neurons

*  When the number of neurons in hidden layer is increased fromn 2 to 3 the error increases to **147.45**
*   The above result shows that error does not always reduce with increase in number of neurons, error can only be reduced by using back propagation





